### PR TITLE
apps/examples/testcase: Testcase for compression using LZMA algorithm

### PR DESCRIPTION
**Pull request by Nishtha97**
- Checking if the data is correctly retrieved after performing compression and
  decompression on a buffer.

Signed-off-by: Nishtha97 <n.maloo@samsung.com>

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [41b93ec90ee247569084f21807e94877caae4dd6] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [41b93ec90ee247569084f21807e94877caae4dd6] - Code Rule Check Result:
apps/examples/testcase/tc_main.c:160: ERROR: [IDT_M_TAB] code indent should use tabs where possible
apps/examples/testcase/tc_main.c:160: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/tc_main.c:326: ERROR: [IDT_M_TAB] code indent should use tabs where possible
apps/examples/testcase/tc_main.c:326: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/tc_main.c:327: ERROR: [IDT_M_TAB] code indent should use tabs where possible
apps/examples/testcase/tc_main.c:327: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/tc_main.c:328: ERROR: [IDT_M_TAB] code indent should use tabs where possible
apps/examples/testcase/tc_main.c:328: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/testcase/tc_main.c:329: ERROR: [IDT_M_TAB] code indent should use tabs where possible
apps/examples/testcase/tc_main.c:329: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

File: apps/examples/testcase/le_tc/compression/tc_lzma.c [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+ *   Check and verify the checksum of a file of size 2KB after compression and decompression
+ *
+ * Returned Value:
+ *   None
+ ****************************************************************************/
+void tc_2Kb(void)
+{
+
+	int i, ret;
+	unsigned int ini_t, after_t;
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/compression/tc_lzma.c]: Please separate the declarations for each variables. (coding rule)

File: apps/examples/testcase/le_tc/compression/tc_lzma.c [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	unsigned int writesize = 2048;
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/compression/tc_lzma.c]: How about using define like `#define BUFFER_SIZE 2048`?

File: apps/examples/testcase/le_tc/compression/tc_lzma.c [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	for (i = 0; i < 2048; i++) {
+		arr[i] = 'A' + ((rand()) % 26);
+	}
+
+	ini_t = crc32(arr, 2048);
+
+	ret = LzmaCompress(&out_buf[LZMA_PROPS_SIZE], &writesize, arr, 2048, out_buf, &propsSize, 0, 1 << 13, -1, -1, -1, -1, 1);
+
+	if (ret != SZ_OK) {
+		printf("LZMA Compress failed, ret = %d\n", ret);
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/compression/tc_lzma.c]: All tc should follow the TC_ASSERT_XXX or TC_SUCCESS_RESULT macros.

**Comment by seinfra** [comment]
Target : [bddcf65d03f93d06a3cd8c5fb8a262f72d12ae27] - Code Rule Check (C++) OK.

File: apps/examples/testcase/le_tc/compression/tc_lzma.c (hunk 1 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_lzma.c]: Made the changes

File: apps/examples/testcase/le_tc/compression/tc_lzma.c (hunk 2 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_lzma.c]: We have two cases in the code wherein buffer_size is 2048 in one and 4096 in the other case. Hence declaration is done separately in each function.

File: apps/examples/testcase/le_tc/compression/tc_internal.h [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/// @file tc_internal.h
+
+/// @brief Header file for Kernel TestCase Example
```
Comment (amandeepc) [inline apps/examples/testcase/le_tc/compression/tc_internal.h]: Kernel ->Compression

File: apps/examples/testcase/le_tc/compression/tc_lzma.c [hunk 5]
```diff
@@ -0,0 +1,142 @@
+/****************************************************************************
+ *
+ * Copyright 2016 Samsung Electronics All Rights Reserved.
```
Comment (amandeepc) [inline apps/examples/testcase/le_tc/compression/tc_lzma.c]: 2016 -> 2019

File: apps/examples/testcase/le_tc/compression/compression_tc_main.c [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+/// @file compression_tc_main.c
+
+/// @brief Main Function for Compression TestCase Example
+
+#include <tinyara/config.h>
+#include <stdio.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <tinyara/testcase_drv.h>
```
Comment (amandeepc) [inline apps/examples/testcase/le_tc/compression/compression_tc_main.c]: Remove header files that are not needed

File: apps/examples/testcase/Kconfig [hunk 7]
```diff
@@ -20,6 +20,7 @@ source "$APPSDIR/examples/testcase/le_tc/kernel/Kconfig"
 source "$APPSDIR/examples/testcase/le_tc/network/Kconfig"
 source "$APPSDIR/examples/testcase/le_tc/ttrace/Kconfig"
 source "$APPSDIR/examples/testcase/le_tc/tcp_tls/Kconfig"
+source "$APPSDIR/examples/testcase/le_tc/compression/Kconfig"
```
Comment (sunghan-chang) [inline apps/examples/testcase/Kconfig]: Please add it by Alphabet order in le_tc tab

File: apps/examples/testcase/Makefile [hunk 8]
```diff
@@ -158,6 +158,9 @@ endif
 ifeq ($(CONFIG_EXAMPLES_TESTCASE_KERNEL),y)
 	$(Q) $(call REGISTER,kernel_tc,tc_kernel_main,TASH_EXECMD_ASYNC,180,2048)
 endif
+ifeq ($(CONFIG_EXAMPLES_TESTCASE_COMPRESSION),y)
```
Comment (sunghan-chang) [inline apps/examples/testcase/Makefile]: same as previous

File: apps/examples/testcase/le_tc/compression/tc_internal.h [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+#ifndef __EXAMPLES_TESTCASE_COMPRESSION_TC_INTERNAL_H
+#define __EXAMPLES_TESTCASE_COMPRESSION_TC_INTERNAL_H
+
+#include "tc_common.h"
+
+#ifdef CONFIG_CPP_HAVE_VARARGS
+
+#ifdef CONFIG_DEBUG_TC_KN
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/compression/tc_internal.h]: Those are not unnecessary

File: apps/examples/testcase/le_tc/compression/tc_lzma.c [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+/****************************************************************************
+ * Name: tc_2Kb
+ *
+ * Description:
+ *   Check and verify the checksum of a file of size 2KB after compression and decompression
+ *
+ * Returned Value:
+ *   None
+ ****************************************************************************/
+void tc_2Kb(void)
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/compression/tc_lzma.c]:
```
static void tc_compression_lzma_2kb(void)
```

File: apps/examples/testcase/le_tc/compression/tc_lzma.c [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+	ret = LzmaUncompress(&arr[0], &max, &out_buf[LZMA_PROPS_SIZE], &writesize, &out_buf[0], LZMA_PROPS_SIZE);
+
+	if (ret == SZ_ERROR_FAIL) {
+		bmdbg("Failure to decompress with LZMAUncompress API\n");
+		ret = -ret;
+	}
+
+	after_t = crc32(arr, 2048);
+
+	if (ini_t == after_t) {
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/compression/tc_lzma.c]: You should use TC_ASSERT_XX macro. Every tc function must have a TC_ASSERT macro at least.

File: apps/examples/testcase/le_tc/compression/tc_lzma.c (hunk 10 above)
Comment (sunghan-chang) [reply apps/examples/testcase/le_tc/compression/tc_lzma.c]: TC naming rule is `<prefix>_<module name>_<test name>_<postfix>`.
prefix is `tc`, `utc` or `itc`.
postfix is `p` or `n`.

File: apps/examples/testcase/le_tc/compression/tc_lzma.c (hunk 3 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_lzma.c]: Done

File: apps/examples/testcase/le_tc/compression/tc_lzma.c (hunk 5 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_lzma.c]: Done

File: apps/examples/testcase/Kconfig (hunk 7 above)
Comment (Nishtha97) [reply apps/examples/testcase/Kconfig]: Done

File: apps/examples/testcase/le_tc/compression/tc_internal.h (hunk 9 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_internal.h]: Made the changes

File: apps/examples/testcase/le_tc/compression/tc_lzma.c (hunk 10 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_lzma.c]: Made the changes

**Comment by seinfra** [comment]
Target : [ebda2e239fc029ba0a3d1ee9fa499fc4aa99a216] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ebda2e239fc029ba0a3d1ee9fa499fc4aa99a216] - Code Rule Check OK.

File: apps/examples/testcase/le_tc/compression/tc_lzma.c (hunk 11 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_lzma.c]: Made the changes

File: apps/examples/testcase/Makefile (hunk 8 above)
Comment (sunghan-chang) [reply apps/examples/testcase/Makefile]: could you check here? This is not listed by Alphabet order.

File: apps/examples/testcase/le_tc/compression/compression_tc_main.c [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+/// @brief Main Function for Compression TestCase Example
+
+#include <tinyara/config.h>
+#include <stdio.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <tinyara/fs/fs.h>
+#include <tinyara/fs/ioctl.h>
+#include "tc_common.h"
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/compression/compression_tc_main.c]: Almost headers are unnecessary for this file. Please clean up

File: apps/examples/testcase/le_tc/compression/tc_internal.h [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ ****************************************************************************/
+
+/// @file tc_internal.h
+
+/// @brief Header file for Compression TestCase Example
+#ifndef __EXAMPLES_TESTCASE_COMPRESSION_TC_INTERNAL_H
+#define __EXAMPLES_TESTCASE_COMPRESSION_TC_INTERNAL_H
+
+#include "tc_common.h"
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/compression/tc_internal.h]: unnecessary

File: apps/examples/testcase/le_tc/compression/compression_tc_main.c (hunk 6 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/compression_tc_main.c]: Done

File: apps/examples/testcase/le_tc/compression/tc_internal.h (hunk 13 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_internal.h]: Made the changes

File: apps/examples/testcase/le_tc/compression/compression_tc_main.c (hunk 12 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/compression_tc_main.c]: Done

File: apps/examples/testcase/Makefile (hunk 8 above)
Comment (Nishtha97) [reply apps/examples/testcase/Makefile]: Done

**Comment by seinfra** [comment]
Target : [acde3ff0dc03329e2f76e2556bd869fb45c8e012] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [acde3ff0dc03329e2f76e2556bd869fb45c8e012] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [46c59c96623f4c26527a142d8a38c3ece385226f] - Code Rule Check (C++) OK.

File: apps/examples/testcase/le_tc/compression/tc_lzma.c [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+	unsigned int writesize = 2048;
+	unsigned int destLen = 2048;
+	unsigned char arr[2048];
+	unsigned char out_buf[2048 + LZMA_PROPS_SIZE];
+
+	for (i = 0; i < 2048; i++) {
+		arr[i] = 'A' + ((rand()) % 26);
+	}
+
+	/* CRC of the file initially */
```
[the remark this answers is not in the document]
Comment (sunghan-chang) [reply apps/examples/testcase/le_tc/compression/tc_lzma.c]: @Nishtha97 Could you check @amandeepc 's comment? Let's merge this by EOD.

**Comment by seinfra** [comment]
Target : [fce51701f3e07308af5b970206ab742ede09f7a4] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fce51701f3e07308af5b970206ab742ede09f7a4] - Code Rule Check OK.

File: apps/examples/testcase/le_tc/compression/tc_lzma.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_lzma.c]: Done

File: apps/examples/testcase/le_tc/compression/tc_lzma.c (hunk 2 above)
Comment (Nishtha97) [reply apps/examples/testcase/le_tc/compression/tc_lzma.c]: Made the changes

**Comment by seinfra** [comment]
Target : [285924ddd7397dca46477ca0e0c668d1a8dd2e9a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [285924ddd7397dca46477ca0e0c668d1a8dd2e9a] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [58b51f7032180859ae7a062762bf63dd3aceb6dd] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [58b51f7032180859ae7a062762bf63dd3aceb6dd] - Code Rule Check OK.

**Comment by sunghan-chang** [comment]
@Nishtha97 Lastly, could you let me know that this is for os/compression or for external/lzma?

**Pull request merged**